### ISSUE TITLE
`<functional>`: better has function

### ISSUE THREAD

**Issue by AlexGuteniev** [microsoft/STL#2360]
@lhecker observes FNV1a is bad for integers.

https://discord.com/channels/737189251069771789/737189251497721887/913600700846596147

**vNext** note: Resolving this issue will require breaking binary compatibility. We won't be able to accept pull requests for this issue until the vNext branch is available. See #169 for more information.

**Comment by AlexGuteniev** [comment]
https://github.com/microsoft/STL/pull/686#issuecomment-609261700 mentions SipHash-2-4

**Comment by lhecker** [comment]
I did some benchmarks based on https://github.com/veorq/SipHash.
The test involves putting numbers 0 to 127 into a `std::unordered_map`, choosing 128 random keys and calling `find()` for at least 100ms for each of those and computing the average time per random key. The table below contains the mean/median/stddev between those 128 randomly chosen keys.

Benchmark           |        Time |           CPU 
--------------------|------------:|--------------:
BM_fnv1_mean        |    1.910 ns |      1.910 ns 
BM_fnv1_median      |    1.910 ns |      1.740 ns 
BM_fnv1_stddev      |    0.014 ns |      0.186 ns 
BM_murmur_mean      |    1.890 ns |      1.880 ns 
BM_murmur_median    |    1.710 ns |      1.740 ns 
BM_murmur_stddev    |    0.323 ns |      0.343 ns 
BM_xxh3_mean        |    2.550 ns |      2.550 ns 
BM_xxh3_median      |    2.300 ns |      2.340 ns 
BM_xxh3_stddev      |    0.333 ns |      0.393 ns 
BM_siphash24_mean   |   18.300 ns |     18.300 ns 
BM_siphash24_median |   17.800 ns |     17.400 ns 
BM_siphash24_stddev |    0.735 ns |      1.640 ns 

(AMD 5950X, SMT, CCD 1 only, fixed 4.2GHz)

Considering things like VSO-653642 it might be difficult to introduce SipHash into a language like C++.

**Comment by StephanTLavavej** [comment]
Thanks - the SipHash perf data is definitely news to me, so we'll need to consider the available options carefully.

**Comment by cbezault** [comment]
I would consider also looking at SipHash13 https://github.com/rust-lang/rust/issues/29754#issuecomment-156073946

**Comment by AlexGuteniev** [comment]
@BillyONeal also mentioned a security consideration:
> the hash values are not the same in successive runs of the program to make unordered_Xxx safe to use with untrusted input out of the box.

But I'm not sure if this can be implemented with insignificant perf impact.

**Comment by BillyONeal** [comment]
I would argue that the default, user doesn't know any better, option, should be the not exploitable one, even at a performance cost. Moreover, the resulting strength of the hash may help some programs by reducing collisions vs. other hash functions; although it will not for very small keys in relatively small containers like 127 int32s.

If you want the absolute fastest of juggling razor blades fast, `unordered_map` is not your port of call in the first place.

**Comment by AlexGuteniev** [comment]
> If you want the absolute fastest of juggling razor blades fast, `unordered_map` is not your port of call in the first place.

`std::hash` is not neccessarily for `unordered_map`, a proper hash table may use it either

**Comment by lhecker** [comment]
@cbezault For short inputs the performance difference between SipHash-2-4 and 1-3 is minimal. I've updated the table above.
@BillyONeal Good point. I agree that a better, but slower hash function should be preferred as the default one for the stdlib. However SipHash isn't a stronger hash function than alternatives and it's rate of collisions is the same as for any other one that more or less passes the smhasher suite.

Personally I'm against using SipHash as the default for this STL _so far_. I believe the absolute majority of code written with this STL are end-user, desktop applications and not publicly exposed web servers. In fact I'd explicitly argue that the number of web servers written in C++ _and_ running on Windows is quite low. While the safety promises of SipHash are nice, it doesn't excuse the increase in computing power consumption for applications that don't require any such safety at all.

Before using a new hash function I believe there's a large number of alternatives we should first consider.
[smhasher](https://github.com/rurban/smhasher) has useful reference code, results and benchmarks on lower end hardware, which could be used as a point of reference for us.

**Comment by StephanTLavavej** [comment]
I agree that we'll need to explore the increasingly large space of alternatives (when vNext work resumes). On the STL Discord, I observed that C++'s concerns are somewhat different from Python's; notably, we don't use dictionaries as our universal data structure (so the impact of a hash function choice is *dramatically* smaller), and `map` is available as an alternative to `unordered_map` and although `map` is typically somewhat slower, it also offers absolute immunity against hash-denial-of-service attacks.
